Doc/c-api/function.rst

```diff
@@ -173,8 +173,15 @@ There are a few functions specific to Python functions.
    runtime behavior depending on optimization decisions, it does not change
    the semantics of the Python code being executed.
 
-   If the callback returns with an exception set, it must return ``-1``; this
-   exception will be printed as an unraisable exception using
-   :c:func:`PyErr_WriteUnraisable`. Otherwise it should return ``0``.
+   If *event* is ``PyFunction_EVENT_DESTROY``, there may already be a pending
```

Comment:
Does it make sense to guarantee that the other event handlers are never called with an exception set?

Is it forbidden or problematic for the callback to store the pending exception, call API functions, and restore it before returning 0? (The wording seems to imply so.)

Reply:
I don't want to make this a guarantee for the callbacks, because it's quite expensive to check `PyErr_Occurred()` and then save/restore the exception. The intended use case for these callbacks is performance-sensitive, and generally should not involve the callback doing anything that even uses the CPython C API; just doing some cheap updates to internal bookkeeping of e.g. a JIT.

But you're right that if a callback really wants to save/clear/restore the exception itself and do something with the C API while it's cleared, that should be fine. I'll update the wording accordingly.